File: src/test/java/com/example/retailstore/factory/DiscountStrategyFactoryTest.java
package com.example.retailstore.factory;

import com.example.retailstore.model.Bill;
import com.example.retailstore.strategy.*;
import org.junit.jupiter.api.Test;

import static com.example.retailstore.model.UserType.*;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class DiscountStrategyFactoryTest {

    private final DiscountStrategyFactory factory  = new DiscountStrategyFactory();


    @Test
    public void shouldReturnsEmployeeDiscountStrategy() {
        Bill bill = new Bill(1000, false, EMPLOYEE);

        DiscountStrategy strategy = factory.getDiscountStrategy(bill);

        assertTrue(strategy instanceof EmployeeDiscountStrategy);
    }

    @Test
    public void shouldReturnsAffiliateDiscountStrategy() {
        Bill bill = new Bill(1000, false, AFFILIATE);

        DiscountStrategy strategy = factory.getDiscountStrategy(bill);

        assertTrue(strategy instanceof AffiliateDiscountStrategy);
    }

    @Test
    public void shouldReturnsCustomerDiscountStrategy() {
        Bill bill = new Bill(1000, false, CUSTOMER_FOR_TWO_YEARS);

        DiscountStrategy strategy = factory.getDiscountStrategy(bill);

        assertTrue(strategy instanceof CustomerDiscountStrategy);
    }

    @Test
    public void shouldReturnsNoDiscountStrategy() {
        Bill bill = new Bill(1000, true, CUSTOMER_FOR_TWO_YEARS);

        DiscountStrategy strategy = factory.getDiscountStrategy(bill);

        assertTrue(strategy instanceof NoDiscountStrategy);
    }
}

File: src/test/java/com/example/retailstore/factory/DiscountStrategyFactoryTest.kt
package com.example.retailstore.factory

import com.example.retailstore.model.Bill
import com.example.retailstore.model.UserType
import com.example.retailstore.strategy.AffiliateDiscountStrategy
import com.example.retailstore.strategy.CustomerDiscountStrategy
import com.example.retailstore.strategy.EmployeeDiscountStrategy
import com.example.retailstore.strategy.NoDiscountStrategy
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test


class DiscountStrategyFactoryTest {
    private val factory = DiscountStrategyFactory()
    @Test
    fun shouldReturnsEmployeeDiscountStrategy() {
        val bill = Bill(1000.0, false, UserType.EMPLOYEE)

        val strategy = factory.getDiscountStrategy(bill)

        assertTrue(strategy is EmployeeDiscountStrategy)
    }

    @Test
    fun shouldReturnsAffiliateDiscountStrategy() {
        val bill = Bill(1000.0, false, UserType.AFFILIATE)

        val strategy = factory.getDiscountStrategy(bill)

        assertTrue(strategy is AffiliateDiscountStrategy)
    }

    @Test
    fun shouldReturnsCustomerDiscountStrategy() {
        val bill = Bill(1000.0, false, UserType.CUSTOMER_FOR_TWO_YEARS)

        val strategy = factory.getDiscountStrategy(bill)

        assertTrue(strategy is CustomerDiscountStrategy)
    }

    @Test
    fun shouldReturnsNoDiscountStrategy() {
        val bill = Bill(1000.0, true, UserType.CUSTOMER_FOR_TWO_YEARS)

        val strategy = factory.getDiscountStrategy(bill)

        assertTrue(strategy is NoDiscountStrategy)
    }
}
